# Task 2

To solve the following exercises you need the following two data files:

- **charts.csv**: contains information on the Top 200 Spotify charts per country and week: for instance a song identifier, the position in the ranking, and the number of streams (for a given country and week)
- **songs.csv** contains details on individual songs: e.g. the song name, artist, genre, its duration, and more variables

## Exercise 1 (10 points)
- Read in the data sets `songs.csv`. 
- Display the first five observations.
- Print the number of rows and columns. 

In [66]:
import numpy as np
import pandas as pd

In [67]:
df_songs = pd.read_csv('data/songs.csv')
print('First 5 rows are below:')
df_songs.head()

First 5 rows are below:


song_id      artist_name                song_name  genre  \
0  00mBzIWv5gHOYxwuEJXjOG  December Avenue    Sa Ngalan Ng Pag-Ibig    pop   
1  00tnXAltVLyIWok9sCWdjL      Dani Mocanu          Afaceri Ilegale  other   
2  00x2d2mKSaUrCNTNxH7CYN            Sigma                     ALYX  other   
3  011Xn3Wqii67s4aaPgheiu         Marteria  Love, Peace & Happiness    rap   
4  014DA3BdnmD3kI5pBogH7c         Chillies            Cứ Chill Thôi  indie   

   duration_ms  explicit  danceability  loudness  speechiness  acousticness  \
0     285138.0     False         0.518    -3.907       0.0305        0.0851   
1     176248.0     False         0.588    -4.680       0.3570        0.4000   
2     167619.0      True         0.495   -10.033       0.3690        0.2800   
3     213701.0     False         0.845    -7.397       0.1470        0.3040   
4     270000.0     False         0.781    -8.452       0.0408        0.2170   

   instrumentalness valence  
0          0.000000   0.271  
1          0.000000   0.425  
2          0.000000   0.368  
3          0.000021   0.377  
4          0.000036   0.609

In [68]:
print('Number of rows are: ', df_songs.shape[0], 'and number of columns are: ', df_songs.shape[1])

Number of rows are:  7388 and number of columns are:  12


## Exercise 2 (10 points)
Display the columns `song_name`, `artist_name`, and `danceability` of the 10 songs with the lowest danceability.

In [69]:
result = df_songs.nsmallest(10, 'danceability')
result[['song_name', 'artist_name', 'danceability']]

song_name           artist_name  \
3504                                       Jólastjarnan          Memfismafían   
5566  Intro to Fake News - from "Spider-Man: No Way ...     Michael Giacchino   
1706                                       O Helga Natt             Nils Bech   
484                               Can You Feel My Heart  Bring Me The Horizon   
30               Have Yourself A Merry Little Christmas          Judy Garland   
3428                                           Jolevise         Odd Nordstoga   
3647                               No Child Left Behind            Kanye West   
1772                                       Silent Night          Mariah Carey   
120                            Nu tändas tusen juleljus                Carola   
2220                              Sydämeeni joulun teen      Vesa-Matti Loiri   

      danceability  
3504         0.141  
5566         0.143  
1706         0.145  
484          0.150  
30           0.156  
3428         0.160  
3647         0.161  
1772         0.172  
120          0.173  
2220         0.175

## Exercise 3 (10 points)
- The column `duration_ms` represents the song duration in milliseconds. Create a new column `duration_min` that represents the duration in minutes. 
- Display all songs longer than or equal to 15 minutes.

In [70]:
df_songs['duration_min'] = df_songs['duration_ms']/1000/60
df_songs.head()

song_id      artist_name                song_name  genre  \
0  00mBzIWv5gHOYxwuEJXjOG  December Avenue    Sa Ngalan Ng Pag-Ibig    pop   
1  00tnXAltVLyIWok9sCWdjL      Dani Mocanu          Afaceri Ilegale  other   
2  00x2d2mKSaUrCNTNxH7CYN            Sigma                     ALYX  other   
3  011Xn3Wqii67s4aaPgheiu         Marteria  Love, Peace & Happiness    rap   
4  014DA3BdnmD3kI5pBogH7c         Chillies            Cứ Chill Thôi  indie   

   duration_ms  explicit  danceability  loudness  speechiness  acousticness  \
0     285138.0     False         0.518    -3.907       0.0305        0.0851   
1     176248.0     False         0.588    -4.680       0.3570        0.4000   
2     167619.0      True         0.495   -10.033       0.3690        0.2800   
3     213701.0     False         0.845    -7.397       0.1470        0.3040   
4     270000.0     False         0.781    -8.452       0.0408        0.2170   

   instrumentalness valence  duration_min  
0          0.000000   0.271      4.752300  
1          0.000000   0.425      2.937467  
2          0.000000   0.368      2.793650  
3          0.000021   0.377      3.561683  
4          0.000036   0.609      4.500000

In [71]:
df_songs[df_songs['duration_min'] >= 15]

song_id  artist_name                      song_name  \
1474  1YAi1uIjqSVMWdvsuK4oUQ  Kelvin Momo                     Ivy League   
3506  3fq0e1TpOQFBDi0vuYrePo  Fer Palacio  Previa y Cachengue 40 - Remix   
5284  5WIK8KuxlEWbrQcRDEpHXV    Sun Diego            Apocalyptic Endgame   

        genre  duration_ms  explicit  danceability  loudness  speechiness  \
1474    other     996089.0     False         0.792   -17.201        0.046   
3506    other     928269.0     False         0.595    -7.228        0.217   
5284  hip hop    1300000.0      True         0.454    -4.520        0.275   

      acousticness  instrumentalness valence  duration_min  
1474        0.0146             0.593   0.178     16.601483  
3506        0.1630             0.000   0.646     15.471150  
5284        0.0185             0.000   0.418     21.666667

## Exercise 4 (10 points)
- For each `genre` calculate the minimum, average, and maximum `danceability`, as well as the number of songs that belong to this genre.

In [72]:
df_songs.head()

song_id      artist_name                song_name  genre  \
0  00mBzIWv5gHOYxwuEJXjOG  December Avenue    Sa Ngalan Ng Pag-Ibig    pop   
1  00tnXAltVLyIWok9sCWdjL      Dani Mocanu          Afaceri Ilegale  other   
2  00x2d2mKSaUrCNTNxH7CYN            Sigma                     ALYX  other   
3  011Xn3Wqii67s4aaPgheiu         Marteria  Love, Peace & Happiness    rap   
4  014DA3BdnmD3kI5pBogH7c         Chillies            Cứ Chill Thôi  indie   

   duration_ms  explicit  danceability  loudness  speechiness  acousticness  \
0     285138.0     False         0.518    -3.907       0.0305        0.0851   
1     176248.0     False         0.588    -4.680       0.3570        0.4000   
2     167619.0      True         0.495   -10.033       0.3690        0.2800   
3     213701.0     False         0.845    -7.397       0.1470        0.3040   
4     270000.0     False         0.781    -8.452       0.0408        0.2170   

   instrumentalness valence  duration_min  
0          0.000000   0.271      4.752300  
1          0.000000   0.425      2.937467  
2          0.000000   0.368      2.793650  
3          0.000021   0.377      3.561683  
4          0.000036   0.609      4.500000

In [73]:
result = df_songs.groupby('genre').agg(
    minimum_d =('danceability', 'min'),
    average_d =('danceability', 'mean'),
    maximum_d =('danceability', 'max'), 
    num_songs = ('song_id', 'count')
)

result

minimum_d  average_d  maximum_d  num_songs
genre                                                        
alternative            0.303   0.625437      0.797         32
dance                  0.394   0.659350      0.843         40
electronic             0.262   0.682172      0.964        232
folk                   0.160   0.614743      0.942         74
hip hop                0.161   0.718825      0.975       1156
house                  0.367   0.696558      0.951         77
indie                  0.250   0.605300      0.980        210
japanese               0.310   0.631611      0.927        149
korean                 0.362   0.652156      0.961         77
latin                  0.342   0.715427      0.943        171
other                  0.141   0.666230      0.974       2540
pop                    0.145   0.616312      0.968       1173
rap                    0.337   0.735078      0.979        651
reggaeton              0.400   0.753335      0.942        212
rock                   0.176   0.600646      0.952        263
singer songwriter      0.316   0.609429      0.872         77
soul                   0.269   0.624078      0.962        103
trap                   0.405   0.742344      0.934        151

## Exercise 5 (10 points)
Calculate the average `valence` of all songs. (10 points)

In [74]:
df_songs.head()

song_id      artist_name                song_name  genre  \
0  00mBzIWv5gHOYxwuEJXjOG  December Avenue    Sa Ngalan Ng Pag-Ibig    pop   
1  00tnXAltVLyIWok9sCWdjL      Dani Mocanu          Afaceri Ilegale  other   
2  00x2d2mKSaUrCNTNxH7CYN            Sigma                     ALYX  other   
3  011Xn3Wqii67s4aaPgheiu         Marteria  Love, Peace & Happiness    rap   
4  014DA3BdnmD3kI5pBogH7c         Chillies            Cứ Chill Thôi  indie   

   duration_ms  explicit  danceability  loudness  speechiness  acousticness  \
0     285138.0     False         0.518    -3.907       0.0305        0.0851   
1     176248.0     False         0.588    -4.680       0.3570        0.4000   
2     167619.0      True         0.495   -10.033       0.3690        0.2800   
3     213701.0     False         0.845    -7.397       0.1470        0.3040   
4     270000.0     False         0.781    -8.452       0.0408        0.2170   

   instrumentalness valence  duration_min  
0          0.000000   0.271      4.752300  
1          0.000000   0.425      2.937467  
2          0.000000   0.368      2.793650  
3          0.000021   0.377      3.561683  
4          0.000036   0.609      4.500000

In [75]:
df_songs['valence'] = pd.to_numeric(df_songs['valence'], errors='coerce')

# Calculate the average valence, ignoring NaN values
average_valence = df_songs['valence'].mean()

# Display the result
print(f"The average valence of all songs is: {average_valence}")

The average valence of all songs is: 0.521122799891687


## Exercise 6 (20 points) 
For the following tasks you need to merge the two data sets (charts and songs). Specifically:

- First, read in the file `charts.csv` and display its first five columns. You are allowed to rename columns if you consider it helpful.
- Second, analyse potential merge problems: 1) How many songs from charts.csv have no corresponding song metadata? 2) How many songs from songs.csv do not appear in the charts data set. State which merge type(s) is/are appropriate, if your goal is to loose no information from any of the two datasets.
- Third, carry out the merge operation, and print the shape of the resulting data frame.

In [79]:
charts_df = pd.read_csv('data/charts.csv', sep=';')
charts_df.head()

CHART_ID      TYPE COUNTRY INTERVAL        DATE  \
0  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
1  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
2  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
3  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
4  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   

                  SONG_ID  POSITION  STREAMS  
0  5PjdY0CKGZdEuoNab3yDmX       1.0    63427  
1  5Z9KJZvQzH6PFmb8SNkxuk       2.0    56828  
2  2Xr1dTzJee307rmrkt8c0g       3.0    55401  
3  02MWAaffLxlfxAUY7c5dvx       4.0    53732  
4  46IZ0fSY2mpAiktS3KOqds       5.0    50906

In [80]:
charts_df.rename(columns={"SONG_ID":"song_id"}, inplace=True)
charts_df.head()

CHART_ID      TYPE COUNTRY INTERVAL        DATE  \
0  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
1  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
2  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
3  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   
4  regional-ae-weekly-2021-12-23  regional      ae   weekly  2021-12-23   

                  song_id  POSITION  STREAMS  
0  5PjdY0CKGZdEuoNab3yDmX       1.0    63427  
1  5Z9KJZvQzH6PFmb8SNkxuk       2.0    56828  
2  2Xr1dTzJee307rmrkt8c0g       3.0    55401  
3  02MWAaffLxlfxAUY7c5dvx       4.0    53732  
4  46IZ0fSY2mpAiktS3KOqds       5.0    50906

In [90]:
#To analyze, Let's merge first using outer join
df_outermerged = pd.merge(charts_df, df_songs, on='song_id', how='outer')

# Rows containing only information from the chatts_df data set
only_charts = df_outermerged[df_outermerged['CHART_ID'].notna() & df_outermerged['artist_name'].isna()].shape[0]

# # Rows containing only information from the df_songs data set
only_songs = df_outermerged[df_outermerged['artist_name'].notna() & df_outermerged['CHART_ID'].isna()].shape[0]


# Print results

print("Number of rows containing only information from the CHARTS data set with NO Corrosponding song metadata:", only_charts)
print("Number of rows containing only information from the SONGS data set with No Corrosponding Charts metadata:", only_songs)


Number of rows containing only information from the CHARTS data set with NO Corrosponding song metadata: 53
Number of rows containing only information from the SONGS data set with No Corrosponding Charts metadata: 376


Based on the results, 53 rows in the charts.csv dataset have no corresponding metadata in the songs.csv dataset, while 376 rows in songs.csv do not appear in charts.csv. This indicates potential discrepancies between the two datasets, such as missing or unmatched song_id values. To ensure no information is lost, an outer merge is the most suitable approach, as it retains all rows from both datasets, filling missing values with NaN where necessary. This approach allows for comprehensive analysis while identifying and handling unmatched data effectively.

In [91]:
df_outermerged = pd.merge(charts_df, df_songs, on='song_id', how='outer')
df_outermerged.shape

(24371, 20)

## Exercise 7 (15 points)

Consider the charts of the week prior to christmas (date: 2021-12-23). How many songs that appeared in the German (de) charts during this week also appear in the charts of the Unites States (us)?

In [92]:
df_outermerged.head()

CHART_ID      TYPE COUNTRY INTERVAL        DATE  \
0  regional-au-weekly-2021-12-23  regional      au   weekly  2021-12-23   
1  regional-de-weekly-2021-04-08  regional      de   weekly  2021-04-08   
2  regional-de-weekly-2021-11-18  regional      de   weekly  2021-11-18   
3  regional-gb-weekly-2021-12-23  regional      gb   weekly  2021-12-23   
4  regional-ie-weekly-2021-12-23  regional      ie   weekly  2021-12-23   

                  song_id  POSITION   STREAMS artist_name song_name genre  \
0  003vvx7Niy0yvhvHt4a68B      50.0  532410.0         NaN       NaN   NaN   
1  003vvx7Niy0yvhvHt4a68B     186.0  379778.0         NaN       NaN   NaN   
2  003vvx7Niy0yvhvHt4a68B     191.0  382759.0         NaN       NaN   NaN   
3  003vvx7Niy0yvhvHt4a68B      77.0  785222.0         NaN       NaN   NaN   
4  003vvx7Niy0yvhvHt4a68B     104.0   77144.0         NaN       NaN   NaN   

   duration_ms explicit  danceability  loudness  speechiness  acousticness  \
0          NaN      NaN           NaN       NaN          NaN           NaN   
1          NaN      NaN           NaN       NaN          NaN           NaN   
2          NaN      NaN           NaN       NaN          NaN           NaN   
3          NaN      NaN           NaN       NaN          NaN           NaN   
4          NaN      NaN           NaN       NaN          NaN           NaN   

   instrumentalness  valence  duration_min  
0               NaN      NaN           NaN  
1               NaN      NaN           NaN  
2               NaN      NaN           NaN  
3               NaN      NaN           NaN  
4               NaN      NaN           NaN

In [104]:
german = df_outermerged[(df_outermerged['DATE'] <= '2021-12-23') & (df_outermerged['COUNTRY'] == 'de')]
usa = df_outermerged[(df_outermerged['DATE'] <= '2021-12-23') & (df_outermerged['COUNTRY'] == 'us')]

# Find the intersection of song IDs between the German and US charts
common_songs = set(german['song_id']).intersection(set(usa['song_id']))

# Count the number of common songs
num_common_songs = len(common_songs)

# Print the result
print(f"Number of songs in the German charts that also appear in the US charts on 2021-12-23: {num_common_songs}")

Number of songs in the German charts that also appear in the US charts on 2021-12-23: 111


In [105]:
# Filter the DataFrame for the week prior to Christmas (2021-12-23) and German charts
german_charts = df_outermerged[(df_outermerged['DATE'] == '2021-12-23') & (df_outermerged['COUNTRY'] == 'de')]

# Filter the DataFrame for the week prior to Christmas (2021-12-23) and US charts
us_charts = df_outermerged[(df_outermerged['DATE'] == '2021-12-23') & (df_outermerged['COUNTRY'] == 'us')]

# Merge the two DataFrames on 'song_id' to find common songs
common_songs = pd.merge(german_charts, us_charts, on='song_id', how='inner')

# Count the number of common songs
num_common_songs = common_songs['song_id'].nunique()

# Print the result
print(f"Number of songs in the German charts that also appear in the US charts on 2021-12-23: {num_common_songs}")


Number of songs in the German charts that also appear in the US charts on 2021-12-23: 64


## Exercise 8 (15 points)

The variable `chart_id` contains information on several different aspects, which are currently also represented via the 4 variables `type`, `country`, `interval` and `date`. Your task is to parse `chart_id` into its components. Specifically:

- Drop the columns `type`, `country`, `interval` and `date` permanently from the data set.
- Then recover these four variables by parsing the string column `chart_id` appropriately.
- Report the column `chart_id` and the four new columns for the first five rows of the data set.

In [106]:
df_outermerged.head()

CHART_ID      TYPE COUNTRY INTERVAL        DATE  \
0  regional-au-weekly-2021-12-23  regional      au   weekly  2021-12-23   
1  regional-de-weekly-2021-04-08  regional      de   weekly  2021-04-08   
2  regional-de-weekly-2021-11-18  regional      de   weekly  2021-11-18   
3  regional-gb-weekly-2021-12-23  regional      gb   weekly  2021-12-23   
4  regional-ie-weekly-2021-12-23  regional      ie   weekly  2021-12-23   

                  song_id  POSITION   STREAMS artist_name song_name genre  \
0  003vvx7Niy0yvhvHt4a68B      50.0  532410.0         NaN       NaN   NaN   
1  003vvx7Niy0yvhvHt4a68B     186.0  379778.0         NaN       NaN   NaN   
2  003vvx7Niy0yvhvHt4a68B     191.0  382759.0         NaN       NaN   NaN   
3  003vvx7Niy0yvhvHt4a68B      77.0  785222.0         NaN       NaN   NaN   
4  003vvx7Niy0yvhvHt4a68B     104.0   77144.0         NaN       NaN   NaN   

   duration_ms explicit  danceability  loudness  speechiness  acousticness  \
0          NaN      NaN           NaN       NaN          NaN           NaN   
1          NaN      NaN           NaN       NaN          NaN           NaN   
2          NaN      NaN           NaN       NaN          NaN           NaN   
3          NaN      NaN           NaN       NaN          NaN           NaN   
4          NaN      NaN           NaN       NaN          NaN           NaN   

   instrumentalness  valence  duration_min  
0               NaN      NaN           NaN  
1               NaN      NaN           NaN  
2               NaN      NaN           NaN  
3               NaN      NaN           NaN  
4               NaN      NaN           NaN

In [107]:
df_outermerged = df_outermerged.drop(columns= ['TYPE', 'COUNTRY', 'INTERVAL', 'DATE'], axis=1)

In [110]:
split_names = df_outermerged['CHART_ID'].str.split("-", n=6, expand=True)
df_outermerged['TYPE'] = split_names[0]
df_outermerged['COUNTRY'] = split_names[1]
df_outermerged['INTERVAL'] = split_names[2]
df_outermerged['DATE'] = split_names[3] + "-" + split_names[4] + '-' + split_names[5] 

df_outermerged.head()

CHART_ID                 song_id  POSITION   STREAMS  \
0  regional-au-weekly-2021-12-23  003vvx7Niy0yvhvHt4a68B      50.0  532410.0   
1  regional-de-weekly-2021-04-08  003vvx7Niy0yvhvHt4a68B     186.0  379778.0   
2  regional-de-weekly-2021-11-18  003vvx7Niy0yvhvHt4a68B     191.0  382759.0   
3  regional-gb-weekly-2021-12-23  003vvx7Niy0yvhvHt4a68B      77.0  785222.0   
4  regional-ie-weekly-2021-12-23  003vvx7Niy0yvhvHt4a68B     104.0   77144.0   

  artist_name song_name genre  duration_ms explicit  danceability  loudness  \
0         NaN       NaN   NaN          NaN      NaN           NaN       NaN   
1         NaN       NaN   NaN          NaN      NaN           NaN       NaN   
2         NaN       NaN   NaN          NaN      NaN           NaN       NaN   
3         NaN       NaN   NaN          NaN      NaN           NaN       NaN   
4         NaN       NaN   NaN          NaN      NaN           NaN       NaN   

   speechiness  acousticness  instrumentalness  valence  duration_min  \
0          NaN           NaN               NaN      NaN           NaN   
1          NaN           NaN               NaN      NaN           NaN   
2          NaN           NaN               NaN      NaN           NaN   
3          NaN           NaN               NaN      NaN           NaN   
4          NaN           NaN               NaN      NaN           NaN   

       TYPE COUNTRY INTERVAL        DATE  
0  regional      au   weekly  2021-12-23  
1  regional      de   weekly  2021-04-08  
2  regional      de   weekly  2021-11-18  
3  regional      gb   weekly  2021-12-23  
4  regional      ie   weekly  2021-12-23

## Bonus (5 points)

How many different artists have a song with a `song_name` that contains the word "Christmas"?

In [115]:
num = df_outermerged['song_name'].str.contains('CHRISTMAS', case=False, na=False).nunique('artist_name')
num

2